[Tracing SDK][4] Add pyproject.toml (#15434)
mlflow/mlflow: packages/tracing/README.md

> # MLflow Trace: An Open-Source SDK for Observability and Monitoring GenAI Applications🔍

Reviewer: Can we use `MLflow Tracing` for consitency?

[![Latest Docs](https://img.shields.io/badge/docs-latest-success.svg?style=for-the-badge)](https://mlflow.org/docs/latest/index.html)
[![Apache 2 License](https://img.shields.io/badge/license-Apache%202-brightgreen.svg?style=for-the-badge&logo=apache)](https://github.com/mlflow/mlflow/blob/master/LICENSE.txt)
[![Slack](https://img.shields.io/badge/slack-@mlflow--users-CF0E5B.svg?logo=slack&logoColor=white&labelColor=3F0E40&style=for-the-badge)](https://mlflow.org/community/#slack)
[![Twitter](https://img.shields.io/twitter/follow/MLflow?style=for-the-badge&labelColor=00ACEE&logo=twitter&logoColor=white)](https://twitter.com/MLflow)

![Trace Hero](https://mlflow.org/docs/latest/assets/images/tracing-top-dcca046565ab33be6afe0447dd328c22.gif)

> MLflow Trace is an open-source, lightweight Python package that only includes the minimum set of dependencies and functionality

Reviewer 2: Please add the pypi package name after referring to `MLflow Trace`

to instrument your code/models/agents with [MLflow Tracing](https://mlflow.org/docs/latest/tracing). This package is designed to be

> used in environments where you want to minimize the size of your dependencies, such as in production or serverless environments.

Reviewer 2: This reads a bit vague. I'd want to minimize my install footprint in edge-deployed models and in a lambda architecture or in certain real-time deployment services (keeping my docker container as small as possible), but the term 'production' or 'serverless' is a bit misleading for why this package is going to exist.

Author: I agree serverless is a vague term here. I think "production" is not misleading because the benefit of reducing dependencies is not only about container start-up but also long-term dependency management. The proposal also said `To instrument their production GenAI deployments with tracing, while minimizing dependencies, ...`

## ✨ Features

- [Automatic Tracing](https://mlflow.org/docs/latest/tracing/integrations/) for AI libraries (OpenAI, LangChain, DSPy, Anthropic, etc...). Follow the link for the full list of supported libraries.
- [Manual instrumentation APIs](https://mlflow.org/docs/latest/tracing/api/manual-instrumentation) such as `@trace` decorator.
- [Production Monitoring](https://mlflow.org/docs/latest/tracing/production)
- Other tracing APIs such as `mlflow.set_trace_tag`, `mlflow.search_traces`, etc.

> - Authentication

Reviewer 2: Can we be more specific about 'authentication'? Providing authenticated access to traces-at-rest / authenticated access for logging trace events

Author: This is about how to authenticate tracing client to log traces to backend, e.g. setting DATABRICKS_HOST. I think we actually don't need to write this out here.

## 🌐 Choose Backend

> The MLflow Trace package is designed to work with te remote hosted MLflow server as a backend. This allows you to log your traces to a central location, making it easier to manage and analyze your traces. There are several different options for hosting your MLflow server, including:

Reviewer 3: ```suggestion
The MLflow Trace package is designed to work with a remote hosted MLflow server as a backend. This allows you to log your traces to a central location, making it easier to manage and analyze your traces. There are several different options for hosting your MLflow server, including:
```

- [Databricks](https://docs.databricks.com/machine-learning/mlflow/managed-mlflow.html) - Databricks offers a FREE, fully managed MLflow server as a part of their platform. This is the easiest way to get started with MLflow tracing, without having to set up any infrastructure.

> - [Amazon SageMaker](https://aws.amazon.com/sagemaker-ai/experiments/) - MLflow on Amazon SageMaker is a fully managed service offer by AWS, including tracing and other MLflow features such as model registry.

Reviewer 2: ```suggestion
- [Amazon SageMaker](https://aws.amazon.com/sagemaker-ai/experiments/) - MLflow on Amazon SageMaker is a fully managed service offered as part of the SageMaker platform by AWS, including tracing and other MLflow features such as model registry.
```

- [Self-hosting](https://mlflow.org/docs/latest/tracking/#tracking_setup) - MLflow is a fully open-source project, allowing you to self-host your own MLflow server and keep your data private. This is a great option if you want to have full control over your data and infrastructure.

## 🚀 Getting Started

### Installation

To install the MLflow Python package, run the following command:

```bash
pip install mlflow-tracing
```

### Connect to the MLflow Server

To connect to your MLflow server to log your traces, set the `MLFLOW_TRACKING_URI` environment variable or use the `mlflow.set_tracking_uri` function:

```python
import mlflow

mlflow.set_tracking_uri("databricks")
# Specify the experiment to log the traces to
mlflow.set_experiment("/Path/To/Experiment")
```

### Start Logging Traces

```python
import openai

client = openai.OpenAI(api_key="<your-api-key>")

# Enable auto-tracing for OpenAI
mlflow.openai.autolog()

# Call the OpenAI API as usual
response = client.chat.completions.create(
    model="gpt-o1-mini",
    messages=[{"role": "user", "content": "Hello, how are you?"}],
)
```

## 📘 Documentation

> Official documentation for MLflow can be found at [here](https://mlflow.org/docs/latest/index.html).

Reviewer 3: should this link to the tracing docs instead? i wonder if people will feel intimidated landing on the mlflow main docs page

Author: Agreed!

![Trace Hero](https://mlflow.org/docs/latest/assets/images/tracing-top-dcca046565ab33be6afe0447dd328c22.gif)

MLflow Tracing (`mlflow-tracing`) is an open-source, lightweight Python package that only includes the minimum set of dependencies and functionality
to instrument your code/models/agents with [MLflow Tracing Feature](https://mlflow.org/docs/latest/tracing). It is designed to be a perfect fit for production environments where you want:

- **⚡️ Faster Deployment**: The package size and dependencies are significantly smaller than the full MLflow package, allowing for faster deployment times in dynamic environments such as Docker containers, serverless functions, and cloud-based applications.
- **🔧 Simplified Dependency Management**: A smaller set of dependencies means less work keeping up with dependency updates, security patches, and breaking changes from upstream libraries.
- **📦 Portability**: With the less number of dependencies, MLflow Tracing can be easily deployed across different environments and platforms, without worrying about compatibility issues.

> - **🔒 Less Security Risks**: Each dependency potentially introduces security vulnerabilities. By reducing the number of dependencies, MLflow Tracing minimizes the attack surface and reduces the risk of security breaches.

Reviewer: Suggestion: The phrase "Less Security Risks" should be changed to "Fewer Security Risks" for grammatical correctness, since "risks" is a countable noun. (Comment by Copilot on behalf of harupy)

- [Databricks](https://docs.databricks.com/machine-learning/mlflow/managed-mlflow.html) - Databricks offers a FREE, fully managed MLflow server as a part of their platform. This is the easiest way to get started with MLflow tracing, without having to set up any infrastructure.
- [Amazon SageMaker](https://aws.amazon.com/sagemaker-ai/experiments/) - MLflow on Amazon SageMaker is a fully managed service offered as part of the SageMaker platform by AWS, including tracing and other MLflow features such as model registry.
- [Nebius](https://nebius.com/) - Nebius, a cutting-edge cloud platform for GenAI explorers, offers a fully managed MLflow server.
- [Self-hosting](https://mlflow.org/docs/latest/tracking/#tracking_setup) - MLflow is a fully open-source project, allowing you to self-host your own MLflow server and keep your data private. This is a great option if you want to have full control over your data and infrastructure.

## 🚀 Getting Started

### Installation

> To install the MLflow Python package, run the following command:

Reviewer: ```suggestion
To install the MLflow Tracing Python package, run the following command:
```

```bash

> pip install mlflow-tracing

Reviewer: can we add a command to install from the source using the github URL? That's useful for developers.

- [Databricks](https://docs.databricks.com/machine-learning/mlflow/managed-mlflow.html) - Databricks offers a FREE, fully managed MLflow server as a part of their platform. This is the easiest way to get started with MLflow tracing, without having to set up any infrastructure.
- [Amazon SageMaker](https://aws.amazon.com/sagemaker-ai/experiments/) - MLflow on Amazon SageMaker is a fully managed service offered as part of the SageMaker platform by AWS, including tracing and other MLflow features such as model registry.
- [Nebius](https://nebius.com/) - Nebius, a cutting-edge cloud platform for GenAI explorers, offers a fully managed MLflow server.
- [Self-hosting](https://mlflow.org/docs/latest/tracking/#tracking_setup) - MLflow is a fully open-source project, allowing you to self-host your own MLflow server and keep your data private. This is a great option if you want to have full control over your data and infrastructure.

## 🚀 Getting Started

> ### Installation

Reviewer: Can we add a note that this package should not be co-installed with mlflow or mlflow-skinny?
